Source/ResponseSerialization.swift

```diff
@@ -144,25 +144,54 @@ extension DataRequest {
-                let end = CFAbsoluteTimeGetCurrent()
-
-                let response = DataResponse(request: self.request,
-                                            response: self.response,
-                                            data: self.data,
-                                            metrics: self.metrics,
-                                            serializationDuration: (end - start),
-                                            result: result)
-
-                self.eventMonitor?.request(self, didParseResponse: response)
-
-                (queue ?? .main).async { completionHandler(response) }
+        appendResponseSerializer {
+            let start = CFAbsoluteTimeGetCurrent()
+            let result = Result { try responseSerializer.serialize(request: self.request,
+                                                                   response: self.response,
+                                                                   data: self.data,
+                                                                   error: self.error) }
+            let end = CFAbsoluteTimeGetCurrent()
+
+            let response = DataResponse(request: self.request,
+                                        response: self.response,
+                                        data: self.data,
+                                        metrics: self.metrics,
+                                        serializationDuration: (end - start),
+                                        result: result)
+
+            self.eventMonitor?.request(self, didParseResponse: response)
+
+            if let serializerError = result.error {
+                self.delegate?.retryRequest(self, dueTo: serializerError) { retryResult in
+                    switch retryResult {
+                    case .doNotRetry:
+                        self.responseSerializerDidComplete {
+                            (queue ?? .main).async { completionHandler(response) }
+                        }
+
+                    case .doNotRetryWithError(let retryError):
+                        let result = Result<Serializer.SerializedObject>.failure(retryError)
+
+                        let response = DataResponse(request: self.request,
+                                                    response: self.response,
+                                                    data: self.data,
+                                                    metrics: self.metrics,
+                                                    serializationDuration: (end - start),
+                                                    result: result)
+
+                        self.responseSerializerDidComplete {
+                            (queue ?? .main).async { completionHandler(response) }
+                        }
+
+                    default:
+                        // No-op: request is being retried
+                        break
+                    }
```

Comment:
I wonder if we should change this so that it's actually up to the caller to execute the retry. As you can see here, it's a bit awkward to rely on the internal implementation that retry has actually fired.

Reply:
I think you're right. Serializer calls and wrapping those calls in retry logic are separate concerns, so they're better separated, if we can.

Reply:
Refactored to be separate in 687c3f1a. So much better in every way.